presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java

```diff
@@ -58,29 +57,23 @@ public static Block toMap(
             ConnectorSession session,
             @SqlType("map(FK,FV)") Block fromMap)
     {
-        // loop over all the parameter types and bind ConnectorSession if needed
-        // TODO: binding `ConnectorSession` should be done in function invocation framework
-        Class<?>[] parameterTypes = keyCastFunction.type().parameterArray();
-        for (int i = 0; i < parameterTypes.length; i++) {
-            if (parameterTypes[i] == ConnectorSession.class) {
-                keyCastFunction = MethodHandles.insertArguments(keyCastFunction, i, session);
-                break;
-            }
-        }
-        parameterTypes = valueCastFunction.type().parameterArray();
-        for (int i = 0; i < parameterTypes.length; i++) {
-            if (parameterTypes[i] == ConnectorSession.class) {
-                valueCastFunction = MethodHandles.insertArguments(valueCastFunction, i, session);
-                break;
-            }
-        }
+        boolean keyCastRequiresSession = keyCastFunction.type().parameterArray()[0] == ConnectorSession.class;
+        boolean valueCastRequiresSession = valueCastFunction.type().parameterArray()[0] == ConnectorSession.class;
 
         TypedSet typedSet = new TypedSet(toKeyType, fromMap.getPositionCount() / 2, "map-to-map cast");
         BlockBuilder keyBlockBuilder = toKeyType.createBlockBuilder(new BlockBuilderStatus(), fromMap.getPositionCount() / 2);
         for (int i = 0; i < fromMap.getPositionCount(); i += 2) {
             Object fromKey = readNativeValue(fromKeyType, fromMap, i);
             try {
-                Object toKey = keyCastFunction.invoke(fromKey);
+                Object toKey;
+                // TODO: binding `ConnectorSession` should be done in function invocation framework
+                if (!keyCastRequiresSession) {
+                    toKey = keyCastFunction.invoke(fromKey);
+                }
+                else {
+                    toKey = keyCastFunction.invoke(session, fromKey);
```

Comment:
This should be done in the specialize method. Take a look at ArrayJoin

Reply:
I.e., there shouldn't be any conditional logic for adapting target handles in the main loop of the function.

Reply:
Actually, let's get this in in the current shape (since it fixes some bad reliability issues) and file an issue to optimize the implementation later.

Reply:
Filed https://github.com/prestodb/presto/issues/9066 for this.